File: NEsper.Tests/regression/event/TestLegacyBeanEvents.cs
using System;
using System.Collections.Generic;

using net.esper.client;
using net.esper.compat;
using net.esper.events;
using net.esper.support.bean;
using net.esper.support.util;

using NUnit.Core;
using NUnit.Framework;

namespace net.esper.regression.events
{
    [TestFixture]
    public class TestLegacyBeanEvents
    {
        private SupportLegacyBean legacyBean;
        private EPServiceProvider epService;

        [SetUp]
        public virtual void setUp()
        {
            EDictionary<String, String> mappedProperty = new HashDictionary<String, String>();
            mappedProperty.Put("key1", "value1");
            mappedProperty.Put("key2", "value2");
            legacyBean = new SupportLegacyBean("leg", new String[] { "a", "b" }, mappedProperty, "nest");
        }

        [TearDown]
        public virtual void tearDown()
        {
            epService = null;
            legacyBean = null;
        }

        [Test]
        public void testPublicAccessors()
        {
            tryPublicAccessors(ConfigurationEventTypeLegacy.CodeGenerationEnum.ENABLED);
        }

        [Test]
        public void testPublicAccessorsNoCodeGen()
        {
            tryPublicAccessors(ConfigurationEventTypeLegacy.CodeGenerationEnum.DISABLED);
        }

        [Test]
        public void testExplicitOnly()
        {
            tryExplicitOnlyAccessors(ConfigurationEventTypeLegacy.CodeGenerationEnum.ENABLED);
        }

        [Test]
        public void testExplicitOnlyNoCodeGen()
        {
            tryExplicitOnlyAccessors(ConfigurationEventTypeLegacy.CodeGenerationEnum.DISABLED);
        }

        [Test]
        public void testJavaBeanAccessor()
        {
            tryJavaBeanAccessor(ConfigurationEventTypeLegacy.CodeGenerationEnum.ENABLED);
        }

        [Test]
        public void testJavaBeanAccessorNoCodeGen()
        {
            tryJavaBeanAccessor(ConfigurationEventTypeLegacy.CodeGenerationEnum.DISABLED);
        }

        [Test]
        public void testFinalClass()
        {
            tryFinalClass(ConfigurationEventTypeLegacy.CodeGenerationEnum.ENABLED);
        }

        [Test]
        public void testFinalClassNoCodeGen()
        {
            tryFinalClass(ConfigurationEventTypeLegacy.CodeGenerationEnum.DISABLED);
        }

        private void tryPublicAccessors(ConfigurationEventTypeLegacy.CodeGenerationEnum codeGeneration)
        {
            Configuration config = new Configuration();

            ConfigurationEventTypeLegacy legacyDef = new ConfigurationEventTypeLegacy();
            legacyDef.AccessorStyle = ConfigurationEventTypeLegacy.AccessorStyleEnum.PUBLIC;
            legacyDef.CodeGeneration = codeGeneration;
            legacyDef.AddFieldProperty("explicitFSimple", "fieldLegacyVal");
            legacyDef.AddFieldProperty("explicitFIndexed", "fieldStringArray");
            legacyDef.AddFieldProperty("explicitFNested", "fieldNested");
            legacyDef.AddMethodProperty("explicitMSimple", "readLegacyBeanVal");
            legacyDef.AddMethodProperty("explicitMArray", "readStringArray");
            legacyDef.AddMethodProperty("explicitMIndexed", "readStringIndexed");
            legacyDef.AddMethodProperty("explicitMMapped", "readMapByKey");
            config.AddEventTypeAlias("MyLegacyEvent", typeof(SupportLegacyBean).FullName, legacyDef);

            legacyDef = new ConfigurationEventTypeLegacy();
            legacyDef.AccessorStyle = ConfigurationEventTypeLegacy.AccessorStyleEnum.PUBLIC;
            legacyDef.CodeGeneration = ConfigurationEventTypeLegacy.CodeGenerationEnum.DISABLED;
            config.AddEventTypeAlias("MyLegacyNestedEvent", typeof(SupportLegacyBean.LegacyNested).FullName, legacyDef);

            epService = EPServiceProviderManager.GetProvider(this.GetType().Name + ".test1" + codeGeneration, config);
            epService.Initialize();

            String statementText =
                "select " + 
                "fieldLegacyVal as fieldSimple," + 
                "fieldStringArray as fieldArr," + 
                "fieldStringArray[1] as fieldArrIndexed," + 
                "fieldMapped as fieldMap," + 
                "fieldNested as fieldNested," + 
                "fieldNested.readNestedValue as fieldNestedVal," + 
                "readLegacyBeanVal as simple," + 
                "readLegacyNested as nestedObject," + 
                "readLegacyNested.readNestedValue as nested," +
                "readStringArray[0] as array," +
                "readStringIndexed[1] as indexed," + 
                "readMapByKey('key1') as mapped," +
                "readMap as mapItself," +
                "explicitFSimple, " + 
                "explicitFIndexed[0], " + 
                "explicitFNested, " +
                "explicitMSimple, " + 
                "explicitMArray[0], " +
                "explicitMIndexed[1], " +
                "explicitMMapped('key2')" +
                " from MyLegacyEvent.win:length(5)";

            EPStatement statement = epService.EPAdministrator.CreateEQL(statementText);
            SupportUpdateListener listener = new SupportUpdateListener();
            statement.AddListener(listener);

            EventType eventType = statement.EventType;
            Assert.AreEqual(typeof(String), eventType.GetPropertyType("fieldSimple"));
            Assert.AreEqual(typeof(String[]), eventType.GetPropertyType("fieldArr"));
            Assert.AreEqual(typeof(String), eventType.GetPropertyType("fieldArrIndexed"));
            Assert.AreEqual(typeof(IDictionary<String, String>), eventType.GetPropertyType("fieldMap"));
            Assert.AreEqual(typeof(SupportLegacyBean.LegacyNested), eventType.GetPropertyType("fieldNested"));
            Assert.AreEqual(typeof(String), eventType.GetPropertyType("fieldNestedVal"));
            Assert.AreEqual(typeof(String), eventType.GetPropertyType("simple"));
            Assert.AreEqual(typeof(SupportLegacyBean.LegacyNested), eventType.GetPropertyType("nestedObject"));
            Assert.AreEqual(typeof(String), eventType.GetPropertyType("nested"));
            Assert.AreEqual(typeof(String), eventType.GetPropertyType("array"));
            Assert.AreEqual(typeof(String), eventType.GetPropertyType("indexed"));
            Assert.AreEqual(typeof(String), eventType.GetPropertyType("mapped"));
            Assert.AreEqual(typeof(String), eventType.GetPropertyType("explicitFSimple"));
            Assert.AreEqual(typeof(String), eventType.GetPropertyType("explicitFIndexed[0]"));
            Assert.AreEqual(typeof(SupportLegacyBean.LegacyNested), eventType.GetPropertyType("explicitFNested"));
            Assert.AreEqual(typeof(String), eventType.GetPropertyType("explicitMSimple"));
            Assert.AreEqual(typeof(String), eventType.GetPropertyType("explicitMArray[0]"));
            Assert.AreEqual(typeof(String), eventType.GetPropertyType("explicitMIndexed[1]"));
            Assert.AreEqual(typeof(String), eventType.GetPropertyType("explicitMMapped('key2')"));

            epService.EPRuntime.SendEvent(legacyBean);

            Assert.AreEqual(legacyBean.fieldLegacyVal, listener.LastNewData[0]["fieldSimple"]);
            Assert.AreEqual(legacyBean.fieldStringArray, listener.LastNewData[0]["fieldArr"]);
            Assert.AreEqual(legacyBean.fieldStringArray[1], listener.LastNewData[0]["fieldArrIndexed"]);
            Assert.AreEqual(legacyBean.fieldMapped, listener.LastNewData[0]["fieldMap"]);
            Assert.AreEqual(legacyBean.fieldNested, listener.LastNewData[0]["fieldNested"]);
            Assert.AreEqual(legacyBean.fieldNested.readNestedValue(), listener.LastNewData[0]["fieldNestedVal"]);

            Assert.AreEqual(legacyBean.readLegacyBeanVal(), listener.LastNewData[0]["simple"]);
            Assert.AreEqual(legacyBean.readLegacyNested(), listener.LastNewData[0]["nestedObject"]);
            Assert.AreEqual(legacyBean.readLegacyNested().readNestedValue(), listener.LastNewData[0]["nested"]);
            Assert.AreEqual(legacyBean.readStringIndexed(0), listener.LastNewData[0]["array"]);
            Assert.AreEqual(legacyBean.readStringIndexed(1), listener.LastNewData[0]["indexed"]);
            Assert.AreEqual(legacyBean.readMapByKey("key1"), listener.LastNewData[0]["mapped"]);
            Assert.AreEqual(legacyBean.readMap(), listener.LastNewData[0]["mapItself"]);

            Assert.AreEqual(legacyBean.readLegacyBeanVal(), listener.LastNewData[0]["explicitFSimple"]);
            Assert.AreEqual(legacyBean.readLegacyBeanVal(), listener.LastNewData[0]["explicitMSimple"]);
            Assert.AreEqual(legacyBean.readLegacyNested(), listener.LastNewData[0]["explicitFNested"]);
            Assert.AreEqual(legacyBean.readStringIndexed(0), listener.LastNewData[0]["explicitFIndexed[0]"]);
            Assert.AreEqual(legacyBean.readStringIndexed(0), listener.LastNewData[0]["explicitMArray[0]"]);
            Assert.AreEqual(legacyBean.readStringIndexed(1), listener.LastNewData[0]["explicitMIndexed[1]"]);
            Assert.AreEqual(legacyBean.readMapByKey("key2"), listener.LastNewData[0]["explicitMMapped('key2')"]);
        }

        private void tryExplicitOnlyAccessors(ConfigurationEventTypeLegacy.CodeGenerationEnum codeGeneration)
        {
            Configuration config = new Configuration();

            ConfigurationEventTypeLegacy legacyDef = new ConfigurationEventTypeLegacy();
            legacyDef.AccessorStyle = ConfigurationEventTypeLegacy.AccessorStyleEnum.EXPLICIT;
            legacyDef.CodeGeneration = codeGeneration;
            legacyDef.AddFieldProperty("explicitFNested", "fieldNested");
            legacyDef.AddMethodProperty("explicitMNested", "readLegacyNested");
            config.AddEventTypeAlias("MyLegacyEvent", typeof(SupportLegacyBean).FullName, legacyDef);

            legacyDef = new ConfigurationEventTypeLegacy();
            legacyDef.AccessorStyle = ConfigurationEventTypeLegacy.AccessorStyleEnum.EXPLICIT;
            legacyDef.CodeGeneration = codeGeneration;
            legacyDef.AddFieldProperty("fieldNestedClassValue", "fieldNestedValue");
            legacyDef.AddMethodProperty("readNestedClassValue", "readNestedValue");
            config.AddEventTypeAlias("MyLegacyNestedEvent", typeof(SupportLegacyBean.LegacyNested).FullName, legacyDef);

            legacyDef = new ConfigurationEventTypeLegacy();
            legacyDef.AccessorStyle = ConfigurationEventTypeLegacy.AccessorStyleEnum.EXPLICIT;
            legacyDef.CodeGeneration = codeGeneration;
            config.AddEventTypeAlias("MySupportBean", typeof(SupportBean).FullName, legacyDef);

            epService = EPServiceProviderManager.GetProvider(this.GetType().Name + ".test2" + codeGeneration, config);
            epService.Initialize();

            String statementText =
                "select " +
                "explicitFNested.fieldNestedClassValue as fnested, " +
                "explicitMNested.readNestedClassValue as mnested" + 
                " from MyLegacyEvent.win:length(5)";

            EPStatement statement = epService.EPAdministrator.CreateEQL(statementText);
            SupportUpdateListener listener = new SupportUpdateListener();
            statement.AddListener(listener);

            EventType eventType = statement.EventType;
            Assert.AreEqual(typeof(String), eventType.GetPropertyType("fnested"));
            Assert.AreEqual(typeof(String), eventType.GetPropertyType("mnested"));

            epService.EPRuntime.SendEvent(legacyBean);

            Assert.AreEqual(legacyBean.fieldNested.readNestedValue(), listener.LastNewData[0]["fnested"]);
            Assert.AreEqual(legacyBean.fieldNested.readNestedValue(), listener.LastNewData[0]["mnested"]);

            try
            {
                // invalid statement, JavaBean-style getters not exposed
                statementText = "select intPrimitive from MySupportBean.win:length(5)";
                epService.EPAdministrator.CreateEQL(statementText);
            }
            catch (EPStatementException ex)
            {
                // expected
            }
        }

        public virtual void tryJavaBeanAccessor(ConfigurationEventTypeLegacy.CodeGenerationEnum codeGeneration)
        {
            Configuration config = new Configuration();

            ConfigurationEventTypeLegacy legacyDef = new ConfigurationEventTypeLegacy();
            legacyDef.AccessorStyle = ConfigurationEventTypeLegacy.AccessorStyleEnum.NATIVE;
            legacyDef.CodeGeneration = codeGeneration;
            legacyDef.AddFieldProperty("explicitFInt", "fieldIntPrimitive");
            legacyDef.AddMethodProperty("explicitMGetInt", "getIntPrimitive");
            legacyDef.AddMethodProperty("explicitMReadInt", "readIntPrimitive");
            config.AddEventTypeAlias("MyLegacyEvent", typeof(SupportLegacyBeanInt).FullName, legacyDef);

            epService = EPServiceProviderManager.GetProvider(this.GetType().Name + ".test3" + codeGeneration, config);
            epService.Initialize();

            String statementText = "select intPrimitive, explicitFInt, explicitMGetInt, explicitMReadInt " + " from MyLegacyEvent.win:length(5)";

            EPStatement statement = epService.EPAdministrator.CreateEQL(statementText);
            SupportUpdateListener listener = new SupportUpdateListener();
            statement.AddListener(listener);
            EventType eventType = statement.EventType;

            SupportLegacyBeanInt _event = new SupportLegacyBeanInt(10);
            epService.EPRuntime.SendEvent(_event);

            string[] _items = new string[] {
                "intPrimitive",
                "explicitFInt",
                "explicitMGetInt",
                "explicitMReadInt"
            };

            foreach (String name in _items)
            {
                Assert.AreEqual(typeof(int), eventType.GetPropertyType(name));
                Assert.AreEqual(10, listener.LastNewData[0][name]);
            }
        }

        private void tryFinalClass(ConfigurationEventTypeLegacy.CodeGenerationEnum codeGeneration)
        {
            Configuration config = new Configuration();
            ConfigurationEventTypeLegacy legacyDef = new ConfigurationEventTypeLegacy();
            legacyDef.AccessorStyle = ConfigurationEventTypeLegacy.AccessorStyleEnum.NATIVE;
            legacyDef.CodeGeneration = codeGeneration;
            config.AddEventTypeAlias("MyFinalEvent", typeof(SupportBeanFinal).FullName, legacyDef);

            epService = EPServiceProviderManager.GetProvider(this.GetType().Name + ".test4" + codeGeneration);
            epService.Initialize();

            String statementText = "select intPrimitive " + "from " + typeof(SupportBeanFinal).FullName + ".win:length(5)";

            EPStatement statement = epService.EPAdministrator.CreateEQL(statementText);
            SupportUpdateListener listener = new SupportUpdateListener();
            statement.AddListener(listener);

            SupportBeanFinal _event = new SupportBeanFinal(10);
            epService.EPRuntime.SendEvent(_event);
            Assert.AreEqual(10, listener.LastNewData[0]["intPrimitive"]);
        }
    }
}

File: trunk/esper/src/test/java/net/esper/regression/event/TestLegacyBeanEvents.java
package net.esper.regression.event;

import net.esper.client.*;
import net.esper.support.bean.SupportLegacyBean;
import net.esper.support.bean.SupportBean;
import net.esper.support.bean.SupportLegacyBeanInt;
import net.esper.support.bean.SupportBeanFinal;
import net.esper.support.util.SupportUpdateListener;
import net.esper.support.client.SupportConfigFactory;
import net.esper.event.EventType;

import java.util.Map;
import java.util.HashMap;

import junit.framework.TestCase;

public class TestLegacyBeanEvents extends TestCase
{
    private SupportLegacyBean legacyBean;
    private EPServiceProvider epService;

    protected void setUp()
    {
        Map<String, String> mappedProperty = new HashMap<String, String>();
        mappedProperty.put("key1", "value1");
        mappedProperty.put("key2", "value2");
        legacyBean = new SupportLegacyBean("leg", new String[] {"a", "b"}, mappedProperty, "nest");
    }

    public void testPublicAccessors()
    {
        tryPublicAccessors(ConfigurationEventTypeLegacy.CodeGeneration.ENABLED);
    }

    public void testPublicAccessorsNoCodeGen()
    {
        tryPublicAccessors(ConfigurationEventTypeLegacy.CodeGeneration.DISABLED);
    }

    public void testExplicitOnly()
    {
        tryExplicitOnlyAccessors(ConfigurationEventTypeLegacy.CodeGeneration.ENABLED);
    }

    public void testExplicitOnlyNoCodeGen()
    {
        tryExplicitOnlyAccessors(ConfigurationEventTypeLegacy.CodeGeneration.DISABLED);
    }

    public void testJavaBeanAccessor()
    {
        tryJavaBeanAccessor(ConfigurationEventTypeLegacy.CodeGeneration.ENABLED);
    }

    public void testJavaBeanAccessorNoCodeGen()
    {
        tryJavaBeanAccessor(ConfigurationEventTypeLegacy.CodeGeneration.DISABLED);
    }

    public void testFinalClass()
    {
        tryFinalClass(ConfigurationEventTypeLegacy.CodeGeneration.ENABLED);
    }

    public void testFinalClassNoCodeGen()
    {
        tryFinalClass(ConfigurationEventTypeLegacy.CodeGeneration.DISABLED);
    }

    private void tryPublicAccessors(ConfigurationEventTypeLegacy.CodeGeneration codeGeneration)
    {
        Configuration config = SupportConfigFactory.getConfiguration();
        config.getEngineDefaults().getThreading().setInternalTimerEnabled(false);

        ConfigurationEventTypeLegacy legacyDef = new ConfigurationEventTypeLegacy();
        legacyDef.setAccessorStyle(ConfigurationEventTypeLegacy.AccessorStyle.PUBLIC);
        legacyDef.setCodeGeneration(codeGeneration);
        legacyDef.addFieldProperty("explicitFSimple", "fieldLegacyVal");
        legacyDef.addFieldProperty("explicitFIndexed", "fieldStringArray");
        legacyDef.addFieldProperty("explicitFNested", "fieldNested");
        legacyDef.addMethodProperty("explicitMSimple", "readLegacyBeanVal");
        legacyDef.addMethodProperty("explicitMArray", "readStringArray");
        legacyDef.addMethodProperty("explicitMIndexed", "readStringIndexed");
        legacyDef.addMethodProperty("explicitMMapped", "readMapByKey");
        config.addEventTypeAlias("MyLegacyEvent", SupportLegacyBean.class.getName(), legacyDef);

        legacyDef = new ConfigurationEventTypeLegacy();
        legacyDef.setAccessorStyle(ConfigurationEventTypeLegacy.AccessorStyle.PUBLIC);
        legacyDef.setCodeGeneration(ConfigurationEventTypeLegacy.CodeGeneration.DISABLED);
        config.addEventTypeAlias("MyLegacyNestedEvent", SupportLegacyBean.LegacyNested.class.getName(), legacyDef);

        epService = EPServiceProviderManager.getProvider(this.getClass().getName() + ".test1" + codeGeneration, config);
        epService.initialize();

        String statementText = "select " +
                    "fieldLegacyVal as fieldSimple," +
                    "fieldStringArray as fieldArr," +
                    "fieldStringArray[1] as fieldArrIndexed," +
                    "fieldMapped as fieldMap," +
                    "fieldNested as fieldNested," +
                    "fieldNested.readNestedValue as fieldNestedVal," +
                    "readLegacyBeanVal as simple," +
                    "readLegacyNested as nestedObject," +
                    "readLegacyNested.readNestedValue as nested," +
                    "readStringArray[0] as array," +
                    "readStringIndexed[1] as indexed," +
                    "readMapByKey('key1') as mapped," +
                    "readMap as mapItself," +
                    "explicitFSimple, " +
                    "explicitFIndexed[0], " +
                    "explicitFNested, " +
                    "explicitMSimple, " +
                    "explicitMArray[0], " +
                    "explicitMIndexed[1], " +
                    "explicitMMapped('key2')" +
                    " from MyLegacyEvent.win:length(5)";

        EPStatement statement = epService.getEPAdministrator().createEQL(statementText);
        SupportUpdateListener listener = new SupportUpdateListener();
        statement.addListener(listener);

        EventType eventType = statement.getEventType();
        assertEquals(String.class, eventType.getPropertyType("fieldSimple"));
        assertEquals(String[].class, eventType.getPropertyType("fieldArr"));
        assertEquals(String.class, eventType.getPropertyType("fieldArrIndexed"));
        assertEquals(Map.class, eventType.getPropertyType("fieldMap"));
        assertEquals(SupportLegacyBean.LegacyNested.class, eventType.getPropertyType("fieldNested"));
        assertEquals(String.class, eventType.getPropertyType("fieldNestedVal"));
        assertEquals(String.class, eventType.getPropertyType("simple"));
        assertEquals(SupportLegacyBean.LegacyNested.class, eventType.getPropertyType("nestedObject"));
        assertEquals(String.class, eventType.getPropertyType("nested"));
        assertEquals(String.class, eventType.getPropertyType("array"));
        assertEquals(String.class, eventType.getPropertyType("indexed"));
        assertEquals(String.class, eventType.getPropertyType("mapped"));
        assertEquals(String.class, eventType.getPropertyType("explicitFSimple"));
        assertEquals(String.class, eventType.getPropertyType("explicitFIndexed[0]"));
        assertEquals(SupportLegacyBean.LegacyNested.class, eventType.getPropertyType("explicitFNested"));
        assertEquals(String.class, eventType.getPropertyType("explicitMSimple"));
        assertEquals(String.class, eventType.getPropertyType("explicitMArray[0]"));
        assertEquals(String.class, eventType.getPropertyType("explicitMIndexed[1]"));
        assertEquals(String.class, eventType.getPropertyType("explicitMMapped('key2')"));

        epService.getEPRuntime().sendEvent(legacyBean);

        assertEquals(legacyBean.fieldLegacyVal, listener.getLastNewData()[0].get("fieldSimple"));
        assertEquals(legacyBean.fieldStringArray, listener.getLastNewData()[0].get("fieldArr"));
        assertEquals(legacyBean.fieldStringArray[1], listener.getLastNewData()[0].get("fieldArrIndexed"));
        assertEquals(legacyBean.fieldMapped, listener.getLastNewData()[0].get("fieldMap"));
        assertEquals(legacyBean.fieldNested, listener.getLastNewData()[0].get("fieldNested"));
        assertEquals(legacyBean.fieldNested.readNestedValue(), listener.getLastNewData()[0].get("fieldNestedVal"));

        assertEquals(legacyBean.readLegacyBeanVal(), listener.getLastNewData()[0].get("simple"));
        assertEquals(legacyBean.readLegacyNested(), listener.getLastNewData()[0].get("nestedObject"));
        assertEquals(legacyBean.readLegacyNested().readNestedValue(), listener.getLastNewData()[0].get("nested"));
        assertEquals(legacyBean.readStringIndexed(0), listener.getLastNewData()[0].get("array"));
        assertEquals(legacyBean.readStringIndexed(1), listener.getLastNewData()[0].get("indexed"));
        assertEquals(legacyBean.readMapByKey("key1"), listener.getLastNewData()[0].get("mapped"));
        assertEquals(legacyBean.readMap(), listener.getLastNewData()[0].get("mapItself"));

        assertEquals(legacyBean.readLegacyBeanVal(), listener.getLastNewData()[0].get("explicitFSimple"));
        assertEquals(legacyBean.readLegacyBeanVal(), listener.getLastNewData()[0].get("explicitMSimple"));
        assertEquals(legacyBean.readLegacyNested(), listener.getLastNewData()[0].get("explicitFNested"));
        assertEquals(legacyBean.readStringIndexed(0), listener.getLastNewData()[0].get("explicitFIndexed[0]"));
        assertEquals(legacyBean.readStringIndexed(0), listener.getLastNewData()[0].get("explicitMArray[0]"));
        assertEquals(legacyBean.readStringIndexed(1), listener.getLastNewData()[0].get("explicitMIndexed[1]"));
        assertEquals(legacyBean.readMapByKey("key2"), listener.getLastNewData()[0].get("explicitMMapped('key2')"));
    }

    private void tryExplicitOnlyAccessors(ConfigurationEventTypeLegacy.CodeGeneration codeGeneration)
    {
        Configuration config = new Configuration();
        config.getEngineDefaults().getThreading().setInternalTimerEnabled(false);

        ConfigurationEventTypeLegacy legacyDef = new ConfigurationEventTypeLegacy();
        legacyDef.setAccessorStyle(ConfigurationEventTypeLegacy.AccessorStyle.EXPLICIT);
        legacyDef.setCodeGeneration(codeGeneration);
        legacyDef.addFieldProperty("explicitFNested", "fieldNested");
        legacyDef.addMethodProperty("explicitMNested", "readLegacyNested");
        config.addEventTypeAlias("MyLegacyEvent", SupportLegacyBean.class.getName(), legacyDef);

        legacyDef = new ConfigurationEventTypeLegacy();
        legacyDef.setAccessorStyle(ConfigurationEventTypeLegacy.AccessorStyle.EXPLICIT);
        legacyDef.setCodeGeneration(codeGeneration);
        legacyDef.addFieldProperty("fieldNestedClassValue", "fieldNestedValue");
        legacyDef.addMethodProperty("readNestedClassValue", "readNestedValue");
        config.addEventTypeAlias("MyLegacyNestedEvent", SupportLegacyBean.LegacyNested.class.getName(), legacyDef);

        legacyDef = new ConfigurationEventTypeLegacy();
        legacyDef.setAccessorStyle(ConfigurationEventTypeLegacy.AccessorStyle.EXPLICIT);
        legacyDef.setCodeGeneration(codeGeneration);
        config.addEventTypeAlias("MySupportBean", SupportBean.class.getName(), legacyDef);

        epService = EPServiceProviderManager.getProvider(this.getClass().getName() + ".test2" + codeGeneration, config);
        epService.initialize();

        String statementText = "select " +
                    "explicitFNested.fieldNestedClassValue as fnested, " +
                    "explicitMNested.readNestedClassValue as mnested" +
                    " from MyLegacyEvent.win:length(5)";

        EPStatement statement = epService.getEPAdministrator().createEQL(statementText);
        SupportUpdateListener listener = new SupportUpdateListener();
        statement.addListener(listener);

        EventType eventType = statement.getEventType();
        assertEquals(String.class, eventType.getPropertyType("fnested"));
        assertEquals(String.class, eventType.getPropertyType("mnested"));

        epService.getEPRuntime().sendEvent(legacyBean);

        assertEquals(legacyBean.fieldNested.readNestedValue(), listener.getLastNewData()[0].get("fnested"));
        assertEquals(legacyBean.fieldNested.readNestedValue(), listener.getLastNewData()[0].get("mnested"));

        try
        {
            // invalid statement, JavaBean-style getters not exposed
            statementText = "select intPrimitive from MySupportBean.win:length(5)";
            epService.getEPAdministrator().createEQL(statementText);
        }
        catch (EPStatementException ex)
        {
            // expected
        }
    }

    public void tryJavaBeanAccessor(ConfigurationEventTypeLegacy.CodeGeneration codeGeneration)
    {
        Configuration config = new Configuration();
        config.getEngineDefaults().getThreading().setInternalTimerEnabled(false);

        ConfigurationEventTypeLegacy legacyDef = new ConfigurationEventTypeLegacy();
        legacyDef.setAccessorStyle(ConfigurationEventTypeLegacy.AccessorStyle.JAVABEAN);
        legacyDef.setCodeGeneration(codeGeneration);
        legacyDef.addFieldProperty("explicitFInt", "fieldIntPrimitive");
        legacyDef.addMethodProperty("explicitMGetInt", "getIntPrimitive");
        legacyDef.addMethodProperty("explicitMReadInt", "readIntPrimitive");
        config.addEventTypeAlias("MyLegacyEvent", SupportLegacyBeanInt.class.getName(), legacyDef);

        epService = EPServiceProviderManager.getProvider(this.getClass().getName() + ".test3" + codeGeneration, config);
        epService.initialize();

        String statementText = "select intPrimitive, explicitFInt, explicitMGetInt, explicitMReadInt " +
                    " from MyLegacyEvent.win:length(5)";

        EPStatement statement = epService.getEPAdministrator().createEQL(statementText);
        SupportUpdateListener listener = new SupportUpdateListener();
        statement.addListener(listener);
        EventType eventType = statement.getEventType();

        SupportLegacyBeanInt event = new SupportLegacyBeanInt(10);
        epService.getEPRuntime().sendEvent(event);

        for (String name : new String[] {"intPrimitive", "explicitFInt", "explicitMGetInt", "explicitMReadInt"})
        {
            assertEquals(int.class, eventType.getPropertyType(name));
            assertEquals(10, listener.getLastNewData()[0].get(name));
        }
    }

    private void tryFinalClass(ConfigurationEventTypeLegacy.CodeGeneration codeGeneration)
    {
        Configuration config = new Configuration();
        ConfigurationEventTypeLegacy legacyDef = new ConfigurationEventTypeLegacy();
        legacyDef.setAccessorStyle(ConfigurationEventTypeLegacy.AccessorStyle.JAVABEAN);
        legacyDef.setCodeGeneration(codeGeneration);
        config.addEventTypeAlias("MyFinalEvent", SupportBeanFinal.class.getName(), legacyDef);

        epService = EPServiceProviderManager.getProvider(this.getClass().getName() + ".test4" + codeGeneration);
        epService.initialize();

        String statementText = "select intPrimitive " +
                    "from " + SupportBeanFinal.class.getName() + ".win:length(5)";

        EPStatement statement = epService.getEPAdministrator().createEQL(statementText);
        SupportUpdateListener listener = new SupportUpdateListener();
        statement.addListener(listener);

        SupportBeanFinal event = new SupportBeanFinal(10);
        epService.getEPRuntime().sendEvent(event);
        assertEquals(10, listener.getLastNewData()[0].get("intPrimitive"));
    }
}
